### DISCUSSION TITLE
Is monkeypatching a coroutine possible?

### DISCUSSION THREAD

**Discussion by heymoonman** [pytest-dev/pytest#10439] [Q&A] [answered]
I have a function within my database handler that is an async function written for a FastAPI endpoint. It simply pulls the latest document from a MongoDB instance.

I'm trying to provide a monkeypatched response from this function where it returns an example document:

`monkeypatch.setattr(MongoDb, 'get_latest_sys_doc', system_document)`

Where 'MongoDb' is the name of the database handler class, and it is properly imported into the test file. 'System_document' is the fixture I wish to have returned.

**The problem** lies in how it is called in the endpoint I am testing:

`sys_doc: dict = await db.get_latest_sys_doc()`

Because it's called with await it's expecting a callable coroutine that it then waits for the response from. It ends up giving this error:

`TypeError: 'dict' object is not callable`

Which is clearly because await is trying to call the async coroutine but is instead acting on the dictionary object I've provided as a mock.

Is there a better way to go about mocking a response for something like this? There is probably room for debate about whether or not this particular function should be async anyway, but I'd quite like a solution with this setup that allows me to PyTest the endpoint in isolation with mocked responses.

I've tried looking for solutions on the Web and haven't found much, nor much in the PyTest discussions or issues. I may just not be understanding monkeypatching fully.

Any suggestions or solutions are appreciated. Thanks.

**Comment by RonnyPfannschmidt** [comment, marked as answer]
Well, put in a async function instead of a sync one?

If you expect a awaitable of dict, return that instead of a dict

**Reply by heymoonman** [reply]
Turns out the problem was with trying to use a fixture as the monkeypatched function. Removing the fixture tag stopped the error I was getting.

Thanks for the help.
